homeassistant/components/elkm1/__init__.py

```diff
@@ -0,0 +1,267 @@
+from homeassistant.core import HomeAssistant, callback  # noqa
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType  # noqa
+
+DOMAIN = "elkm1"
+
+REQUIREMENTS = ['elkm1-lib==0.7.10']
+
+CONF_AREA = 'area'
+CONF_COUNTER = 'counter'
+CONF_KEYPAD = 'keypad'
+CONF_OUTPUT = 'output'
+CONF_SETTING = 'setting'
+CONF_TASK = 'task'
+CONF_THERMOSTAT = 'thermostat'
+CONF_USER = 'user'
+CONF_PANEL = 'panel'
+CONF_PLC = 'plc'  # Not light because HASS complains about this
+CONF_ZONE = 'zone'
+
+CONF_ENABLED = 'enabled'    # True to enable subdomain
+CONF_HIDE = 'hide'
+CONF_SHOW = 'show'
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA_SUBDOMAIN = vol.Schema({
+    vol.Optional(CONF_ENABLED, default=True): cv.boolean,
+    vol.Optional(CONF_INCLUDE): list,
+    vol.Optional(CONF_EXCLUDE): list,
+    vol.Optional(CONF_HIDE): list,
+    vol.Optional(CONF_SHOW): list,
+    })
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_TEMPERATURE_UNIT, default='fahrenheit'):
+            vol.In(['celsius', 'fahrenheit']),
+        vol.Optional(CONF_AREA): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_COUNTER): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_KEYPAD): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_OUTPUT): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_PLC): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_SETTING): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_TASK): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_THERMOSTAT): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_USER): CONFIG_SCHEMA_SUBDOMAIN,
+        vol.Optional(CONF_ZONE): CONFIG_SCHEMA_SUBDOMAIN,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SUPPORTED_DOMAINS = ['sensor', 'switch', 'alarm_control_panel',
+                     'climate', 'light']
+
+
+async def async_setup(hass: HomeAssistant, hass_config: ConfigType) -> bool:
+    """Set up the Elk M1 platform."""
+
+    from elkm1_lib.const import Max
+    from elkm1_lib.message import housecode_to_index
+    import elkm1_lib as elkm1
+
+    configs = {
+        CONF_AREA: Max.AREAS.value,
+        CONF_COUNTER: Max.COUNTERS.value,
+        CONF_KEYPAD: Max.KEYPADS.value,
+        CONF_OUTPUT: Max.OUTPUTS.value,
+        CONF_PANEL: 1,
+        CONF_PLC: Max.LIGHTS.value,
+        CONF_SETTING: Max.SETTINGS.value,
+        CONF_TASK: Max.TASKS.value,
+        CONF_THERMOSTAT: Max.THERMOSTATS.value,
+        CONF_USER: Max.USERS.value,
+        CONF_ZONE: Max.ZONES.value,
+    }
+
+    def parse_value(val, max_):
+        """Parse a value as an int or housecode."""
+        i = int(val) if val.isdigit() else (housecode_to_index(val) + 1)
```

Comment:
I suppose it could be re-implemented as a voluptuous validator, but we'd just be converting the existing function to voluptuous validator format. The special sauce isn't casting it to an `int` (which voluptuous can do for us just fine) but if it's not a number and instead an X10 style house/device code, converting it to an `int` accordingly.

Reply:
Break out `housecode_to_index(val) + 1` into a validator function, `is_housecode`. Then do eg:
```py
schema = vol.Any(vol.Coerce(int), is_housecode)
val = schema(val)
```
Expand the schema as needed before using it.